### PR TITLE
Improve trino-dbt developer experience

### PR DESCRIPTION
**Pull request by mdesmet**
## Overview

- Use changie to automate the changelog
- Use a pre-commit hook to do linting and formatting (black, mypy flake8
- Add contribution guidelines

## Checklist

- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] `README.md` updated and added information about my change

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+## Getting the code
+
+### Installing git
+
+You will need `git` in order to download and modify the `dbt-core` source code. On macOS, the best way to download git is to just install [Xcode](https://developer.apple.com/support/xcode/).
+
+### How to contribute?
+
+If you are not a member of the `starburstdata` GitHub organization, you can contribute to `dbt-trino` by forking the `dbt-trino` repository. For a detailed overview on forking, check out the [GitHub docs on forking](https://help.github.com/en/articles/fork-a-repo). In short, you will need to:
```
Comment (findinpath) [inline CONTRIBUTING.md:11]: Let's leave `starburstdata` out of the picture. The guideline is relevant mainly for the open source community where the members will likely need to do forks.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
+# Contributing to `dbt-trino`
+
+## Getting the code
```
Comment (findinpath) [inline CONTRIBUTING.md:3]: Very good addition!!!

Thank you @mdesmet

File: README.md [hunk 3]
```diff
@@ -343,75 +343,10 @@ When executing a prepared statement with a large number of parameters, you might
...
-The catalog in the dbt profile can be setup through [pytest markers](https://docs.pytest.org/en/7.1.x/example/markers.html#registering-markers), if no marker has been specified the memory catalog is used.
-
-For example if you want to set the dbt profile to connect to the Delta Lake catalog, annotate your test with `@pytest.mark.delta`, (supported markers are `postgresql`, `delta` or `iceberg`).
-
-```
-@pytest.mark.delta
-def test_run_seed_test(self, project):
-  ...
-```
+- Want to report a bug or request a feature? Let us know on [Slack](http://community.getdbt.com/) in the db-presto-trino channel, or open [an issue](https://github.com/starburstdata/dbt-trino/issues/new)
```
Comment (findinpath) [inline README.md:348]: db-presto-trino -> #db-presto-trino

File: .github/workflows/changelog-existence.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+# **why?**
+# Ensure code change gets reflected in the CHANGELOG.
+
+# **when?**
+# This will run for all PRs going into master.  It will
+# run when they are opened, reopened, when any label is added or removed
+# and when new code is pushed to the branch.  The action will then get
+# skipped if the 'Skip Changelog' label is present is any of the labels.
+
+name: Check Changelog Entry
```
Comment (hovaesco) [inline .github/workflows/changelog-existence.yml:21]: What's the flavour of this job? Is it GHA? I cannot see `runs-on` key.

File: .github/workflows/version-bump.yml [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+# A PR will be created with the changes that can be reviewed before committing.
+
+# **why?**
+# This is to aid in releasing dbt-trino and making sure we have updated
+# the version in all places and generated the changelog.
+
+# **when?**
+# This is triggered manually
+
+name: Version Bump
```
Comment (hovaesco) [inline .github/workflows/version-bump.yml:14]: This is so cool, it would be nice to have something similar for bumping Trino and Starburst versions.

File: .github/workflows/version-bump.yml [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        run: |
+          git checkout -b ${{ steps.variables.outputs.BRANCH_NAME }}
+          git push origin ${{ steps.variables.outputs.BRANCH_NAME }}
+          git branch --set-upstream-to=origin/${{ steps.variables.outputs.BRANCH_NAME }} ${{ steps.variables.outputs.BRANCH_NAME }}
+
+      - name: Bump version
+        run: |
+          source env/bin/activate
+          pip install -r dev-requirements.txt
+          env/bin/bumpversion --allow-dirty --new-version ${{ github.event.inputs.version_number }} major
```
Comment (hovaesco) [inline .github/workflows/version-bump.yml:66]: We used to run two commands to bump version of dbt-core and dbt-trino:
```
bumpversion --config-file .bumpversion-dbt.cfg patch --new-version <new-version>
bumpversion --config-file .bumpversion.cfg patch --new-version <new-version> --allow-dirty

File: .github/workflows/version-bump.yml [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+      - name: Create Pull Request
+        uses: peter-evans/create-pull-request@v3
+        with:
+          author: 'Github Build Bot <buildbot@starburstdata.com>'
+          base: ${{github.ref}}
+          title: 'Bumping version to ${{ github.event.inputs.version_number }} and generate changelog'
+          branch: '${{ steps.variables.outputs.BRANCH_NAME }}'
+          labels: |
+            Skip Changelog
```
Comment (hovaesco) [inline .github/workflows/version-bump.yml:107]: nit: EOF

File: .github/workflows/version-bump.yml [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+          author_name: 'Github Build Bot'
+          author_email: 'buildbot@fishtownanalytics.com'
+          message: 'Bumping version to ${{ github.event.inputs.version_number }} and generate CHANGELOG'
+          branch: '${{ steps.variables.outputs.BRANCH_NAME }}'
+          push: 'origin origin/${{ steps.variables.outputs.BRANCH_NAME }}'
+
+      - name: Create Pull Request
+        uses: peter-evans/create-pull-request@v3
+        with:
+          author: 'Github Build Bot <buildbot@starburstdata.com>'
```
Comment (hovaesco) [inline .github/workflows/version-bump.yml:102]: Is it a valid Starburst github user?

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+### Installing git
+
+You will need `git` in order to download and modify the `dbt-core` source code. On macOS, the best way to download git is to just install [Xcode](https://developer.apple.com/support/xcode/).
```
Comment (hovaesco) [inline CONTRIBUTING.md:7]: What do you think about removing this section? I don't think it's very likely that someone who wants to contribute doesn't have git. The comment about installing git using `Xcode` seems not true, and other OS are not listed.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+You don't need to worry about which `dbt-trino` version your change will go into. Just create the changelog entry with `changie`, and open your PR against the `master` branch. 
+
+## Submitting a Pull Request
+
+A `dbt-trino` maintainer will review your PR. They may suggest code revision for style or clarity, or request that you add unit or integration test(s). These are good things! We believe that, with a little bit of help, anyone can contribute high-quality code.
+
+Automated tests run via GitHub Actions. If you're a first-time contributor, all tests (including code checks and unit tests) will require a maintainer to approve. Changes in the `dbt-trino` repository trigger integration tests against Trino and Starburst.
+
+Once all tests are passing and your PR has been approved, a `dbt-trino` maintainer will merge your changes into the active development branch. And that's it! Happy developing :tada:
```
Comment (hovaesco) [inline CONTRIBUTING.md:156]: I love it!

Comment (hovaesco) [reply CONTRIBUTING.md:156]: `the active development branch` --> `the master branch`

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+# Runs integration tests on Trino
+make dbt-trino-tests
+# Runs integration tests on Starburst
+make dbt-starburst-tests
+```
+> These make targets assume you have a local installation of a recent version of [`tox`](https://tox.readthedocs.io/en/latest/) for unit/integration testing and pre-commit for code quality checks,
+> unless you use choose a Docker container to run tests. Run `make help` for more info.
+
+#### `pre-commit`
+[`pre-commit`](https://pre-commit.com) takes care of running all code-checks for formatting and linting. Run `make dev` to install `pre-commit` in your local environment.  Once this is done you can use any of the linter-based make targets as well as a git pre-commit hook that will ensure proper formatting and linting.
```
Comment (hovaesco) [inline CONTRIBUTING.md:109]: There is no `make dev` recipe in Makefile, could you add a recipe or delete it from here?

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+```sh
+# Runs integration tests on Trino
+make dbt-trino-tests
+# Runs integration tests on Starburst
+make dbt-starburst-tests
+```
+> These make targets assume you have a local installation of a recent version of [`tox`](https://tox.readthedocs.io/en/latest/) for unit/integration testing and pre-commit for code quality checks,
+> unless you use choose a Docker container to run tests. Run `make help` for more info.
+
+#### `pre-commit`
```
Comment (hovaesco) [inline CONTRIBUTING.md:108]: It's not used now, do you have plan to open a subsequent PR which adds `pre-commit`

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+These are the tools used in `dbt-trino` development and testing:
+
+- [`tox`](https://tox.readthedocs.io/en/latest/) to manage virtualenvs across python versions. We currently target the latest patch releases for Python 3.7, 3.8, 3.9, and 3.10
+- [`pytest`](https://docs.pytest.org/en/latest/) to define, discover, and run tests
+- [`flake8`](https://flake8.pycqa.org/en/latest/) for code linting
+- [`black`](https://github.com/psf/black) for code formatting
+- [`mypy`](https://mypy.readthedocs.io/en/stable/) for static type checking
+- [`pre-commit`](https://pre-commit.com) to easily run those checks
+- [`changie`](https://changie.dev/) to create changelog entries, without merge conflicts
+- [`make`](https://users.cs.duke.edu/~ola/courses/programming/Makefiles/Makefiles.html) to run multiple setup or test steps in combination. Don't worry too much, nobody _really_ understands how `make` works, and our Makefile aims to be super simple.
```
Comment (hovaesco) [inline CONTRIBUTING.md:34]: Some tools are not used currently, please open issues if we want to implement them. Black, mypy, pre-commit are very welcome.

File: CONTRIBUTING.md (hunk 11 above)
Comment (mdesmet) [reply CONTRIBUTING.md:109]: ok, will do, note that we have not yet added pre-commit or black. But we could start by running the flake linter with pre-commit.

File: CONTRIBUTING.md (hunk 13 above)
Comment (mdesmet) [reply CONTRIBUTING.md:34]: I will introduce them in separate commits.

Comment (mdesmet) [reply CONTRIBUTING.md:34]: ok so black doesn't deal with imports?

**Comment by mdesmet** [comment]
If you don't create a changelog entry you would get a message on your PR:

image

File: .github/workflows/changelog-existence.yml (hunk 4 above)
Comment (mdesmet) [reply .github/workflows/changelog-existence.yml:21]: We reuse the dbt-labs action, it is set there to `ubuntu-latest`

File: .changie.yaml [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    minInt: 1
+  - key: PR
+    label: GitHub Pull Request Number
+    type: int
+    minInt: 1
+
+footerFormat: |
+  {{- $contributorDict := dict }}
+  {{- /* any names added to this list should be all lowercase for later matching purposes */}}
+  {{- $core_team := list "mdesmet" "findinpath" "hovaesco" "dependabot[bot]" }}
```
Comment (mdesmet) [inline .changie.yaml:45]: We shouldn't filter out the collaborators

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 15]
```diff
@@ -0,0 +1,33 @@
+---
+name: Bug report
```
Comment (hovaesco) [inline .github/ISSUE_TEMPLATE/bug_report.md:2]: Where it's being displayed? I cannot find it in your fork while creating an issue.

File: .github/workflows/stale.yml [hunk 16]
```diff
@@ -0,0 +1,15 @@
+name: 'Close stale issues and PRs'
```
Comment (hovaesco) [inline .github/workflows/stale.yml:1]: This is not very helpful, if some issue is being opened longer than 180 days for example due to some missing functionality in Trino engine then we can easily forget about it and it will be closed. I understand that it can be commented or labelled to prevent removing but it's easy to overlook it in github notifications. We don't have so many issues and PRs in dbt-trino.

File: .github/workflows/ci.yml [hunk 17]
```diff
@@ -12,6 +12,20 @@ on:
...
+        uses: actions/checkout@v2
+
+      - name: "Install Python"
+        uses: actions/setup-python@v2
+
+      - name: "Install dev requirements"
+        run: pip install -r dev_requirements.txt
+
+      - name: "Run pre-commit checks"
+        run: pre-commit run --all-files
```
Comment (hovaesco) [inline .github/workflows/ci.yml:28]: 👍

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+- Your code changes do not unexpectedly break other established functionality
+- Your code changes can handle all known edge cases
+- The functionality you're adding will _keep_ working in the future
+
+### Initial setup
+
+To be able to run the tests locally you will need a Trino or Starburst instance.
+
+```sh
+// to start Trino
```
Comment (hovaesco) [inline CONTRIBUTING.md:80]: `#` instead of `//` and the same below

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+
+## Setting up an environment
+
+There are some tools that will be helpful to you in developing locally. While this is the list relevant for `dbt-trino` development, many of these tools are used commonly across open-source python projects.
+
+### Tools
+
+These are the tools used in `dbt-trino` development and testing:
+
+- [`tox`](https://tox.readthedocs.io/en/latest/) to manage virtualenvs across python versions. We currently target the latest patch releases for Python 3.7, 3.8, 3.9, and 3.10
```
Comment (hovaesco) [inline CONTRIBUTING.md:23]: isort missing on the list

**Review by hovaesco — COMMENTED** [review]
Almost good to go

File: .github/workflows/stale.yml (hunk 16 above)
Comment (mdesmet) [reply .github/workflows/stale.yml:1]: ok let's remove that one

File: .github/ISSUE_TEMPLATE/bug_report.md (hunk 15 above)
Comment (mdesmet) [reply .github/ISSUE_TEMPLATE/bug_report.md:2]: I need to merge it to master there.

Comment (mdesmet) [reply .github/ISSUE_TEMPLATE/bug_report.md:2]: See https://github.com/mdesmet/dbt-trino/issues/new/choose

File: .changes/unreleased/Under the Hood-20220810-173529.yaml [hunk 20]
```diff
@@ -0,0 +1,7 @@
+kind: Under the Hood
+body: Improve developer experience (changie, pre-commit, black, ...)
```
[the remark this answers is not in the document]
Comment (mdesmet) [reply .changes/unreleased/Under the Hood-20220810-173529.yaml]: i replaced it with isort, as that one was the only one missing in the list. Others were already present in the current code.

**Review by hovaesco — APPROVED** [review]
Great job!
